meta/meta.go

```diff
@@ -0,0 +1,181 @@
+package meta
+
+import (
+	"strings"
+
+	"pkg.re/essentialkaos/ek.v8/fmtc"
+	"pkg.re/essentialkaos/ek.v8/fsutil"
+	"pkg.re/essentialkaos/ek.v8/jsonutil"
+	"pkg.re/essentialkaos/ek.v8/path"
+)
+
+// ////////////////////////////////////////////////////////////////////////////////// //
+
+type VMetadata struct {
+	*VMetadataRepository
+	StorageURL  string
+	StoragePath string
+}
+
+type VMetadataRepository struct {
+	Name        string              `json:"name"`
+	Description string              `json:"description"`
+	Versions    []*VMetadataVersion `json:"versions"`
+}
+
+type VMetadataVersion struct {
+	Version   string               `json:"version"`
+	Providers []*VMetadataProvider `json:"providers"`
+}
+
+type VMetadataProvider struct {
+	Name         string `json:"Name"`
+	URL          string `json:"url"`
+	Checksum     string `json:"checksum"`
+	ChecksumType string `json:"checksum_type"`
+}
+
+// ////////////////////////////////////////////////////////////////////////////////// //
+
+func getMetaFormat(name string) string {
+	return fmtc.Sprintf("%s.json", name)
+}
+
+//func AreEqualVersions(first string, second string) bool {
+//	return first == second
+//}
+
+//func NotEqualVersions(first string, second string) bool {
+//	return !AreEqualVersions(first, second)
+//}
+
+func (m *VMetadata) IsEmptyMeta() bool {
+	return len(m.Versions) == 0
+}
+
+func (m *VMetadata) BaseMeta() string {
+	return path.Join(m.StoragePath, m.Name)
+}
+
+func (m *VMetadata) DirMeta() string {
+	return path.Join(m.BaseMeta(), "metadata")
+}
+
+func (m *VMetadata) PathMeta() string {
+	return path.Join(m.DirMeta(), getMetaFormat(m.Name))
+}
+
+func (m *VMetadata) CloneMeta() *VMetadata {
+	clone := *m
+	return &clone
```

Comment:
This code doesn't clone (_deep copy_) original struct - https://play.golang.org/p/4_J9g1iYpp
You should use [reflect](https://golang.org/pkg/reflect/) package for this.